README.md
# perl-Mojo-WebSocketProxy


[![Build Status](https://travis-ci.org/binary-com/perl-Mojo-WebSocketProxy.svg?branch=master)](https://travis-ci.org/binary-com/perl-Mojo-WebSocketProxy)
[![codecov](https://codecov.io/gh/binary-com/perl-Mojo-WebSocketProxy/branch/master/graph/badge.svg)](https://codecov.io/gh/binary-com/perl-Mojo-WebSocketProxy)

#### INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

#### NAME

Mojo::WebSocketProxy - WebSocket proxy for JSON-RPC 2.0 server

#### SYNOPSYS

     # lib/your-application.pm

     use base 'Mojolicious';

     sub startup {
         my $self = shift;
         $self->plugin(
             'web_socket_proxy' => {
                 actions => [
                     ['json_key', {some_param => 'some_value'}]
                 ],
                 base_path => '/api',
                 url => 'http://rpc-host.com:8080/',
             }
         );
    }

Or to manually call RPC server:

     # lib/your-application.pm

     use base 'Mojolicious';

     sub startup {
         my $self = shift;
         $self->plugin(
             'web_socket_proxy' => {
                 actions => [
                     [
                         'json_key',
                         {
                             instead_of_forward => sub {
                                 shift->call_rpc({
                                     args => $args,
                                     method => $rpc_method, # it'll call 'http://rpc-host.com:8080/rpc_method'
                                     rpc_response_cb => sub {...}
                                 });
                             }
                         }
                     ]
                 ],
                 base_path => '/api',
                 url => 'http://rpc-host.com:8080/',
             }
         );
    }

#### DESCRIPTION

Using this module you can forward WebSocket-JSON requests to RPC server.

For every message it creates separate hash ref storage, which is available from hooks as $req\_storage.
Request storage have RPC call parameters in $req\_storage->{call\_params}.
It copies message args to $req\_storage->{call\_params}->{args}.
You can use Mojolicious stash to store data between messages in one connection.

#### Proxy responses

The plugin sends websocket messages to client with RPC response data.
If RPC reponse looks like this:

    {status => 1}

It returns simple response like this:

    {$msg_type => 1, msg_type => $msg_type}

If RPC returns something like this:

    {
        response_data => [..],
        status        => 1,
    }

Plugin returns common response like this:

    {
        $msg_type => $rpc_response,
        msg_type  => $msg_type,
    }

You can customize ws porxy response using 'response' hook.

#### Sequence Diagram

![Alt text](https://g.gravizo.com/source/ws_proxy?https%3A%2F%2Fraw.githubusercontent.com%2Fbinary-com%2Fperl-Mojo-WebSocketProxy%2Fmaster%2FREADME.md)
<details> 
<summary></summary>
ws_proxy
@startuml;
title Websocket Proxy

participant Client

Client->Websocket:Initiate connection
Client->Websocket:Send Message

note over Websocket: before_forward
note over Websocket: instead_of_forward

Websocket->Websocket: instead_of_forward does not forward to rpc and returns response back from ws if its valid one

Websocket->Client: send response (only if instead_of_forward)

note over Websocket: before_call

Websocket->RPC: RPC request

note over Websocket: after_forward
note over Websocket: after_dispatch

note over RPC: processing

note over Websocket: before_got_rpc_response

RPC->Websocket: RPC response

note over Websocket: after_got_rpc_response
note over Websocket: success/error
note over Websocket: response
note over Websocket: before_send_api_response
note over Websocket: send

Websocket->Client:send response back

note over Websocket: after_send_api_response

Client->Websocket:Close Websocket connection
@enduml
ws_proxy
</details>

#### Plugin parameters

The plugin understands the following parameters.

##### actions

A pointer to array of action details, which contain stash\_params,
request-response callbacks, other call parameters.

    $self->plugin(
        'web_socket_proxy' => {
            actions => [
                ['action1_json_key', {details_key1 => details_value1}],
                ['action2_json_key']
            ]
        });

##### before\_forward

    before_forward => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Global hook which will run after request is dispatched and before to start preparing RPC call.
It'll run every hook or until any hook returns some non-empty result.
If returns any hash ref then that value will be JSON encoded and send to client,
without forward action to RPC. To call RPC every hook should return empty or undefined value.
It's good place to some validation or subscribe actions.

#### instead\_of\_forward (global)

    instead_of_forward => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Use this hook if you don't want dispatcher to call RPC and want to handle request
in websocket itself. It's not good practice to use it as global hook because if
if you return response from sub passed then it will return same response for each
call. [Read more](#instead-of-forward)

#### before\_call (global)

    before_call => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Global hook which will run just before making rpc call.

##### after\_forward (global)

    after_forward => [sub { my ($c, $result, $req_storage) = @_; ... }, sub {...}]

Global hook which will run after every forwarded RPC call done.
Or even forward action isn't running.
It can view or modify result value from 'before\_forward' hook.
It'll run every hook or until any hook returns some non-empty result.
If returns any hash ref then that value will be JSON encoded and send to client.

##### after\_dispatch (global)

    after_dispatch => [sub { my $c = shift; ... }, sub {...}]

Global hook which will run at the end of request handling.

##### before\_get\_rpc\_response (global)

    before_get_rpc_response => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Global hook which will run when asynchronous RPC call is answered.

##### after\_got\_rpc\_response (global)

    after_got_rpc_response => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Global hook which will run after checked that response exists.

##### before\_send\_api\_response (global)

    before_send_api_response => [sub { my ($c, $req_storage, $api_response) = @_; ... }, sub {...}]

Global hook which will run immediately before send API response.

##### after\_sent\_api\_response (global)

    before_send_api_response => [sub { my ($c, $req_storage) = @_; ... }, sub {...}]

Global hook which will run immediately after sent API response back to client.

##### base\_path

API url for make route.

##### stream\_timeout

See ["timeout" in Mojo::IOLoop::Stream](https://metacpan.org/pod/Mojo::IOLoop::Stream#timeout)

##### max\_connections

See ["max\_connections" in Mojo::IOLoop](https://metacpan.org/pod/Mojo::IOLoop#max_connections)

##### max\_response\_size

Returns error if RPC response size is over value.

##### opened\_connection

Callback for doing something once after connection is opened

##### finish\_connection

Callback for doing something every time when connection is closed.

##### url

RPC host url - store url string or function to set url dynamically for manually RPC calls.
When using forwarded call then url storing in request storage.
You can store url in every action options, or make it at before\_forward hook.

#### Actions options

##### stash\_params

    stash_params => [qw/ stash_key1 stash_key2 /]

Use this if you want to send specified parameters from Mojolicious $c->stash to RPC.
RPC will receive this as part of call params.

You can store RPC response data to Mojolicious stash returning data like this:

    rpc_response => {
        stash => {..} # data to store in Mojolicious stash
        response_key1 => response_value1, # response to API client
        response_key2 => response_value2
    }

##### success

    success => sub { my ($c, $rpc_response) = @_; ... }

Hook which will run if RPC returns success value.

##### error

    error => sub { my ($c, $rpc_response) = @_; ... }

Hook which will run if RPC returns value with error key, e.g.

    { result => { error => { code => 'some_error' } } }

##### response

    response => sub { my ($c, $rpc_response) = @_; ... }

Hook which will run every time when success or error callbacks is running.
It good place to modify API response format.

#### Instead of forward

This hook is generally used if you don't want to forward request to RPC and want
to handle it within websocket itself, for example like send back server time.

Another case where its useful is if you don't want to send response to rpc url
provided in global scope and want to forward to separate RPC service (useful if
you have multiple RPC service to handle different type of request)

```
 instead_of_forward => sub {
     shift->call_rpc({
         url  => 'some other rpc url',
         args => $args,
         method => $rpc_method, # it'll call 'http://rpc-host.com:8080/rpc_method'
         rpc_response_cb => sub {...}
     });
```

#### SEE ALSO

[Mojolicious::Plugin::WebSocketProxy](https://metacpan.org/pod/Mojolicious::Plugin::WebSocketProxy),
[Mojo::WebSocketProxy](https://metacpan.org/pod/Mojo::WebSocketProxy)
[Mojo::WebSocketProxy::CallingEngine](https://metacpan.org/pod/Mojo::WebSocketProxy::CallingEngine),
[Mojo::WebSocketProxy::Dispatcher](https://metacpan.org/pod/Mojo::WebSocketProxy::Dispatcher),
[Mojo::WebSocketProxy::Config](https://metacpan.org/pod/Mojo::WebSocketProxy::Config)
[Mojo::WebSocketProxy::Parser](https://metacpan.org/pod/Mojo::WebSocketProxy::Parser)

#### COPYRIGHT AND LICENSE

Copyright (C) 2016 binary.com


Changes
Revision history for Perl module Mojo::WebSocketProxy

{{$NEXT}}

0.16      2023-12-15 07:48:13+00:00 UTC

    Mojo::WebSocketProxy::Dispatcher changes:
    - Use Log::Any instead of Mojo::Log in Dispatcher;
    - Avoid string interpolation to solve warning due to the uninitialised value of $c->tx->remote_address;
    - Add fallback value for cases when remote_address can't be resolved;

0.15      2023-10-27 14:29:50+08:00 Asia/Shanghai
    -Logger Wrapper added to Proxy for logging context with request 

0.14
    - Move Mojo::WebSocketProxy::Backend::JobAsync and Mojo::WebSocketProxy::Backend::ConsumerGroups out
    - Add support for separating messages to different channels with
      different timeout values for each channel.
    - Increase minimum Perl version from 5.014 to 5.024 for the module.
    - Remove warning line for RPC communication failure.
    - Add documentation about RPC failure callback.

0.13      2019-09-20 01:44:27+00:00 UTC
    - Change sanity error type
    - Add callback for use on RPC communication failure

0.12      2019-08-15 09:57:43+00:00 UTC
    - remove hardcoded config, read from object

0.11      2018-08-15 13:03:25+08:00 Asia/Manila
    - Fix test dependencies (switching from build phase,) adding Path::Tiny.

0.10      2018-08-14 12:40:37+08:00 Asia/Manila
    - Censor messages failing JSON decoding on Dispatcher::open_connection(),
      moving these from error logging to debug logging to prevent leaking
      private information, as well as dropping the WebSocket connection
      entirely.

0.09      2018-08-01 10:09:57+08:00 Asia/Manila
    - Fix character encoding issues on Dispatcher::open_connection() due
      to Mojo::Transaction::WebSockets emitting JSON-decoded `message` events;
      listen on `text` events instead, which pass raw WS_TEXT frames for
      decoding via JSON::MaybeUTF8
    - Relax Mojolicious dependency to >= 7.29

0.08      2018-07-09 10:08:29+08:00 Asia/Manila
    - Fix Changelog

0.07      2018-07-06 13:11:47+08:00 Asia/Manila
    - Add warning with transaction details failure, if available.
    - Don't bother removing args from req storage after SanityCheckFailed.
    - Normalise received JSON via NFC
    - Add binary_frame hook to capture binary uploaded data.

0.06      2017-06-01 03:05:17+00:00 UTC
    This introduces support for callbacks which return Future instances,
    allowing for deferred processing when the before or after hooks need
    to perform I/O before returning a result.

0.05      2017-05-01 06:30:16+00:00 UTC
    - remove Data::UUID

0.04      2017-02-21 02:16:06+00:00 UTC
    - Switch to dzil

0.03 Thu Jun  9 15:50:00 2016
    - Use before send hook to check response size

0.02 Thu Jun  9 09:28:00 2016
    - Minor fixes

0.01 Thu Jun  2 06:02:18 2016
    - First version, released on an unsuspecting world.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.024.
use strict;
use warnings;

use 5.024000;

use ExtUtils::MakeMaker 7.64;

my %WriteMakefileArgs = (
  "ABSTRACT" => "WebSocket proxy for JSON-RPC 2.0 server",
  "AUTHOR" => "binary.com <BINARY\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.64"
  },
  "DISTNAME" => "Mojo-WebSocketProxy",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.024000",
  "NAME" => "Mojo::WebSocketProxy",
  "PREREQ_PM" => {
    "Class::Method::Modifiers" => 0,
    "Data::UUID" => 0,
    "DataDog::DogStatsd::Helper" => "0.05",
    "Encode" => 0,
    "Future" => "0.36",
    "Future::Mojo" => "0.004",
    "Future::Utils" => 0,
    "IO::Async::Loop::Mojo" => 0,
    "JSON::MaybeUTF8" => 0,
    "Log::Any" => 0,
    "Mojo::Base" => 0,
    "Mojo::Redis2" => 0,
    "Mojo::Util" => 0,
    "MojoX::JSON::RPC" => 0,
    "MojoX::JSON::RPC::Client" => 0,
    "Mojolicious" => "7.29",
    "Scalar::Util" => 0,
    "Syntax::Keyword::Try" => "0.04",
    "Unicode::Normalize" => "1.25",
    "constant" => 0,
    "curry" => "1.001",
    "feature" => 0,
    "indirect" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Future" => "0.36",
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "JSON::MaybeXS" => 0,
    "Mojo::IOLoop" => 0,
    "MojoX::JSON::RPC::Service" => 0,
    "Mojolicious" => "7.29",
    "Net::EmptyPort" => 0,
    "Path::Tiny" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Fatal" => 0,
    "Test::MockModule" => 0,
    "Test::MockObject" => 0,
    "Test::Mojo" => 0,
    "Test::More" => "0.98",
    "Test::Simple" => "0.44",
    "Test::TCP" => 0,
    "Test::Warnings" => 0,
    "base" => 0
  },
  "VERSION" => "0.15",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Class::Method::Modifiers" => 0,
  "Data::UUID" => 0,
  "DataDog::DogStatsd::Helper" => "0.05",
  "Encode" => 0,
  "Exporter" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Future" => "0.36",
  "Future::Mojo" => "0.004",
  "Future::Utils" => 0,
  "IO::Async::Loop::Mojo" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeUTF8" => 0,
  "JSON::MaybeXS" => 0,
  "Log::Any" => 0,
  "Mojo::Base" => 0,
  "Mojo::IOLoop" => 0,
  "Mojo::Redis2" => 0,
  "Mojo::Util" => 0,
  "MojoX::JSON::RPC" => 0,
  "MojoX::JSON::RPC::Client" => 0,
  "MojoX::JSON::RPC::Service" => 0,
  "Mojolicious" => "7.29",
  "Net::EmptyPort" => 0,
  "Path::Tiny" => 0,
  "Scalar::Util" => 0,
  "Syntax::Keyword::Try" => "0.04",
  "Test::CheckDeps" => "0.010",
  "Test::Fatal" => 0,
  "Test::MockModule" => 0,
  "Test::MockObject" => 0,
  "Test::Mojo" => 0,
  "Test::More" => "0.98",
  "Test::Simple" => "0.44",
  "Test::TCP" => 0,
  "Test::Warnings" => 0,
  "Unicode::Normalize" => "1.25",
  "base" => 0,
  "constant" => 0,
  "curry" => "1.001",
  "feature" => 0,
  "indirect" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
